cpp/ql/src/Security/CWE/CWE-311/CleartextTransmission.ql

```diff
@@ -31,12 +34,31 @@ abstract class NetworkSendRecv extends FunctionCall {
    * Gets the expression for the buffer to be sent from / received into.
    */
   abstract Expr getDataExpr();
+
+  /**
+   * Holds if the socket used by this call could be a true network socket (or
+   * if no socket is specified). A constant value is likely to indicate standard
+   * input, standard output or a similar non-network socket.
+   */
+  predicate checkSocket() {
```

Comment:
It doesn't seem to be problematic.  I don't think the class `NetworkSendRecv` ever gets very big.

I can add `bindingset[this]` to the predicate if you're still worried?

Reply:
>I can add `bindingset[this]` to the predicate if you're still worried?

I don't think `FromSensitiveConfiguration.isSink` will compile, then. Will it?

Reply:
Good idea.  TODO

Reply:
It doesn't work to do the `checkSocket` code in the constructor as `checkSocket` depends on `getSocketExpr` from the two subclasses, which in turn depends on the constructor.  I haven't found a good solution that doesn't flatten all three classes together or otherwise make everything unnecessarily horrible.

Reply:
> > I can add `bindingset[this]` to the predicate if you're still worried?
>
> I don't think `FromSensitiveConfiguration.isSink` will compile, then. Will it?

This approach does compile.  I'm not exactly sure why, and I'm not convinced the annotation adds anything.

In any case I think `this` is _sufficiently_ bound, for performance purposes, by the union of constructors of the subclasses (i.e. roughly `this.getTarget() instanceof RemoteFlowSinkFunction or RemoteFlowSourceFunction`).

Tuple counts don't look frightening at all (`git_git` locally):
```
[2021-12-14 17:15:05] (189s) Starting to evaluate predicate CleartextTransmission::NetworkSendRecv::checkSocket_dispred#f#antijoin_rhs/1@d08e4807
[2021-12-14 17:15:05] (189s) Tuple counts for CleartextTransmission::NetworkSendRecv::checkSocket_dispred#f#antijoin_rhs/1@d08e4807 after 0ms:
                      10 ~0%        {2} r1 = JOIN CleartextTransmission::NetworkSendRecv#class#f WITH CleartextTransmission::NetworkSendRecv::getSocketExpr_dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.0 'arg0'
                      10 ~0%        {2} r2 = JOIN r1 WITH ASTValueNumbering::globalValueNumber#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'arg0'
                      
                      54 ~0%        {3} r3 = JOIN r2 WITH ASTValueNumbering::globalValueNumber#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, 123, Lhs.1 'arg0'
                      36 ~1698%     {1} r4 = JOIN r3 WITH exprs ON FIRST 2 OUTPUT Lhs.2 'arg0'
                      
                      54 ~0%        {2} r5 = JOIN r2 WITH ASTValueNumbering::globalValueNumber#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'arg0'
                      15 ~87%       {2} r6 = JOIN r5 WITH Access::Access::getTarget_dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'arg0'
                      8  ~166%      {2} r7 = JOIN r6 WITH Variable::Variable::getInitializer_dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'arg0'
                      8  ~166%      {3} r8 = JOIN r7 WITH initialisers ON FIRST 1 OUTPUT Rhs.2, 123, Lhs.1 'arg0'
                      0  ~0%        {1} r9 = JOIN r8 WITH exprs ON FIRST 2 OUTPUT Lhs.2 'arg0'
                      
                      36 ~1698%     {1} r10 = r4 UNION r9
                                    return r10
```

Reply:
This is just the antijoin, though (i.e., the tuples that are _not_ in the output). So the final number of tuples in `checkSocket` will be larger the smaller, `CleartextTransmission::NetworkSendRecv::checkSocket_dispred#f#antijoin_rhs` is.

Reply:
But you're right in that the final output probably isn't very large since it's bounded by the number of function calls. So this is not at all a blocking issue.

Reply:
I've created a branch based on yours [here](https://github.com/MathiasVP/ql/commit/eb8d87e6a6ca37ce5cb1d7044f6fc97a93fd8e76) that moves `checkSocket` into the charpred. The trick is to define another abstract class that makes `getSocketExpr` available in the charpred of `NetworkSendRecv`. If you like that better, feel free to take inspiration from that. But if you prefer this more straightforward (but possibly more scary-looking) version, feel free to keep it like this.

Reply:
LGTM, I'll add it to my branch and then clean up the QLDoc.

Reply:
Done.

Reply:
Thanks for creating that commit, by the way.